phfs.c

```diff
@@ -13,61 +13,362 @@
+
+	case phfs_prot_phoenixd:
+		return "phoenixd";
 
-	if (pdn >= PDN_NB) {
-		handle.h = ERR_ARG;
-		return handle;
+	default:
+		return "";
 	}
+}
+
+
+static int phfs_setProt(phfs_handler_t *ph, const char *prot)
+{
+	size_t sz = plostd_strlen(prot);
+
+	if (prot == NULL || (sz == 3 && plostd_strncmp(prot, "raw", 3) == 0))
+		ph->prot = phfs_prot_raw;
+	else if (sz == 8 && plostd_strncmp(prot, "phoenixd", 8) == 0)
+		ph->prot = phfs_prot_phoenixd;
+	else
+		return ERR_ARG;
+
+	return ERR_NONE;
+}
+
+
+static int phfs_getHandlerId(const char *alias)
+{
+	int i;
+	size_t sz;
+	phfs_handler_t *ph;
 
-	return phfs_handlers[pdn].open(phfs_handlers[pdn].dn, name, flags);
+	for (i = 0; i < phfs_common.hCnt; ++i) {
+		ph = &phfs_common.handlers[i];
+		sz = plostd_strlen(ph->alias);
+		if ((sz == plostd_strlen(alias)) && (plostd_strncmp(alias, ph->alias, sz) == 0))
+			return i;
+	}
+
+	return ERR_ARG;
 }
 
 
-s32 phfs_read(u16 pdn, handle_t handle, addr_t *pos, u8 *buff, u32 len)
+int phfs_regDev(const char *alias, unsigned int major, unsigned int minor, const char *prot)
 {
-	if (pdn >= PDN_NB)
+	size_t sz;
+	dev_handler_t *dh;
+	phfs_handler_t *ph;
+
+	if (alias == NULL)
 		return ERR_ARG;
 
-	return phfs_handlers[pdn].read(phfs_handlers[pdn].dn, handle, pos, buff, len);
+	if (phfs_common.hCnt >= SIZE_PHFS_HANDLERS) {
+		plostd_printf(ATTR_ERROR, "\nToo many devices");
+		return ERR_ARG;
+	}
+
+	if (devs_getHandler(major, minor, &dh) < 0) {
+		plostd_printf(ATTR_ERROR, "\n%d.%d - device does not exist", major, minor);
+		return ERR_ARG;
+	}
+
+	/* Check if alias is already in use */
+	if (phfs_getHandlerId(alias) >= 0) {
+		plostd_printf(ATTR_ERROR, "\n%s - alias is already in use", alias);
+		return ERR_ARG;
+	}
+
+	ph = &phfs_common.handlers[phfs_common.hCnt];
+	if (phfs_setProt(ph, prot) < 0) {
+		plostd_printf(ATTR_ERROR, "\n%s - wrong protocol name\n\t use: \"%s\", \"%s\"", prot,
+		              phfs_getProtName(phfs_prot_raw), phfs_getProtName(phfs_prot_phoenixd));
+		return ERR_ARG;
+	}
+
+	sz = (plostd_strlen(alias) < 7) ? (plostd_strlen(alias) + 1) : 7;
+	hal_memcpy(ph->alias, alias, sz);
+	ph->alias[sz] = '\0';
+
+	ph->h = dh;
+	ph->major = major;
+	ph->minor = minor;
+	phfs_common.hCnt++;
+
+	return ERR_NONE;
+}
+
+
+void phfs_showDevs(void)
+{
+	int i;
+	phfs_handler_t *ph;
+
+	if (phfs_common.hCnt == 0) {
+		plostd_printf(ATTR_ERROR, "\nNone of the devices have been registered\n");
+		return;
+	}
+
+	plostd_printf(ATTR_LOADER, "\nALIAS\tID\tPROTOCOL\n");
+	for (i = 0; i < phfs_common.hCnt; ++i) {
+		ph = &phfs_common.handlers[i];
+		plostd_printf(ATTR_NONE, "%s\t%d.%d\t%s\n", ph->alias, ph->major, ph->minor, phfs_getProtName(ph->prot));
+	}
+}
+
+
+static int phfs_getFileId(const char *alias)
+{
+	int i;
+	size_t sz;
+	phfs_file_t *file;
+
+	for (i = 0; i < phfs_common.fCnt; ++i) {
+		file = &phfs_common.files[i];
+		sz = plostd_strlen(file->alias);
+		if ((sz == plostd_strlen(alias)) && (plostd_strncmp(alias, file->alias, sz) == 0))
```

Comment:
Same as in `phfs_getHandlerId(const char *alias)`

Reply:
Thanks for noticing. Fixed